hypershift-operator/controllers/machineconfigserver/machineconfigserver_controller.go

```diff
@@ -242,14 +219,122 @@ func (r *MachineConfigServerReconciler) Reconcile(ctx context.Context, req ctrl.
 	}
 
 	mcs.Status.Version = releaseImage.Version()
-	mcs.Status.Host = ignitionRoute.Spec.Host
 	if err := r.Status().Update(ctx, mcs); err != nil {
 		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *MachineConfigServerReconciler) reconcileMCSServiceNodePort(ctx context.Context, mcs *hyperv1.MachineConfigServer, mcsService *corev1.Service) (*ctrl.Result, error) {
+	var serviceNodePort int32 = 0
+	var machineConfigServerServiceData corev1.Service
+	if mcs.Spec.IgnitionService.NodePort != nil && mcs.Spec.IgnitionService.NodePort.Port > 0 {
+		serviceNodePort = mcs.Spec.IgnitionService.NodePort.Port
+	}
+	r.Log.Info("Checking for existing service", "serviceName", mcsService.Name, "namespace", mcsService.Namespace)
+	if err := r.Client.Get(ctx, ctrlclient.ObjectKey{Namespace: mcsService.Namespace, Name: mcsService.Name}, &machineConfigServerServiceData); err != nil && !apierrors.IsNotFound(err) {
+		return &ctrl.Result{}, err
+	}
+	if len(machineConfigServerServiceData.Spec.Ports) > 0 && machineConfigServerServiceData.Spec.Ports[0].NodePort > 0 {
+		r.Log.Info("Existing node port exists for service. Preserving it", "nodePort", machineConfigServerServiceData.Spec.Ports[0].NodePort)
+		serviceNodePort = machineConfigServerServiceData.Spec.Ports[0].NodePort
+	}
+	r.Log.Info("Creating MCS Service")
+	_, err := controllerutil.CreateOrUpdate(ctx, r.Client, mcsService, func() error {
```

Comment:
if `serviceNodePort` is explicitly set on the API, this `CreateOrUpdate` fails the second time through with
```
"error": "failed to ensure infrastructure: failed to create Kube API service: failed to create create api server service: Service \"kube-apiserver\" is invalid: spec.ports[0].nodePort: Invalid value: 32001: provided port is already allocated"
```
It is basically `NodePort` `Service` colliding with itself on `Update`.

In my hacky baremetal patch, I just did a `Get` to see if the service exist and, if so, bailed out.  There isn't much that can be changed about a `Service` once it is initially created.
https://github.com/sjenning/hypershift/commit/3d5dc01302078dafc2e55783ee2a3b04fcb36e84

Reply:
my Services block is
```
  services:
  - service: APIServer
    servicePublishingStrategy:
      type: NodePort
      nodePort:
        address: api.example.ocp.variantweb.net
        port: 32001
  - service: VPN
    servicePublishingStrategy:
      type: NodePort
      nodePort:
        address: api.example.ocp.variantweb.net
        port: 32002
  - service: OAuthServer
    servicePublishingStrategy:
      type: Route
```

Reply:
FWIW, I think an idiomatic way to handle a "create-only" reconcile via `CreateOrUpdate` would be to check `.CreationTimestamp.IsZero()` of the incoming resource and simply return early if the resource already exists.

Reply:
I was hoping to be able to reconcile labels over time but maybe if that happened I would need to get more creative and do a patch instead. Thank you for this I'll adjust

Reply:
"sevice selector labels" in case we needed it.